File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n"); 
const N = Number(input[0]);

// 이차원 배열 생성
let graph = new Array(N).fill(null).map(() => Array(N).fill(0));
for (let y in graph) {
    graph[y] = input[Number(y) + 1].split('').map(Number);
}

const dx = [0, 1, 0, -1];
const dy = [-1, 0, 1, 0];

const inRange = (x, y) => {
    return x >= 0 && x < N && y >= 0 && y < N && graph[x][y] === 1;
};

const dfs = (x, y, visited) => {
    visited[x][y] = true;
    let cnt = 1; // 각 영역의 카운트를 새로 초기화
    for (let i = 0; i < 4; i++) {
        let nx = x + dx[i];
        let ny = y + dy[i];
        if (inRange(nx, ny) && !visited[nx][ny]) {
            cnt += dfs(nx, ny, visited); // 재귀 호출 결과를 합산
        }
    }
    return cnt; // 영역의 카운트 반환
};

let counts = [];
let visited = new Array(N).fill(false).map(() => new Array(N).fill(false));
let totalcnt = 0;
for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
        if (!visited[i][j] && graph[i][j] === 1) {
            const cnt = dfs(i, j, visited);
            counts.push(cnt); // 각 영역의 카운트를 push
            totalcnt++;
        }
    }
}

console.log(totalcnt)
console.log(counts.sort((a,b)=>a-b).join('\n'));

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys
input = sys.stdin.readline

N = int(input()) # 지도의 크기
graph = [list(map(int, input().strip())) for _ in range(N)]

sizes = [] # 단지별 집의 수 

# dfs가 적합할 것 같다.
stack = []
visited = [[False]*N for _ in range(N)] 

dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]
def in_range(x, y):
    return 0<=x<N and 0<=y<N 

def dfs(px, py):
    stack.append((px,py))
    visited[py][px] = True
    size = 1

    while stack:
        nx, ny = stack.pop()
        for i in range(4):
            x, y = nx+dx[i], ny+dy[i]
            # 벽에 부딪히는가?
            if not in_range(x, y): continue 
            # 방문한적 있는가?
            if visited[y][x]: continue
            # 집이 존재하는가? 1인가
            if graph[y][x] == 1: 
                # 방문처리 
                visited[y][x] = True
                stack.append((x,y))
                size+=1
    return size



for y in range(N):
    for x in range(N):
        if not visited[y][x] and graph[y][x] == 1:
            sizes.append(dfs(x, y))

print(len(sizes))
print('\n'.join(map(str,sorted(sizes))))


        
